src/aqua/chain/changed_deals.aqua

```diff
@@ -73,16 +72,15 @@ func poll_deal_changes_batch(spell_id: SpellId, chain: ChainInfo, deals: []Joine
         -- Update all deals with new info
         for change <- changes!:
             deal_id = change.deal_info.deal_id
-
             if change.success:
-                deal_log(spell_id, deal_id, ["found a deal changed log on block", change.log!.block_number])
-
-                -- Update app_cid of the deal on the corresponding worker spell
-                update_worker(spell_id, change)
-
-                -- Move block range so that log's block becomes new left boundary
-                log_block = change.log!.block_number
-                store_deal_state(spell_id, deal_id, DealState(left_boundary = log_block))
+                if change.log != nil:
```

Comment:
How could it be that `change.success` is true, but `change.log` is nil?

Reply:
The idea of the fix is the following:
`get_deal_changes` takes the list of deals and returns the list of the same deals. If the deal is updated (logs != nil) then we updated the deal. For all these deals we update the left_boundary to the corresponding values.

The `success` means that the process of retrieving the update isn't broken, and the lack of updates isn't an error

Reply:
So `log == nil` means "there's no update in that range of blocks", correct?

Reply:
yeah

Reply:
If yes, then you gotta move the range, otherwise it will always ask for the same range, where there are no updates

Reply:
well, we move it in the same function, below

Reply:
But you don't do `store_deal_state` with new `left_boundary`, right?

Reply:
I think you should do the same thing on both `log == nil` and `log != nil`:
```
store_deal_state(spell_id, deal_id, DealState(left_boundary = log_block))
```

Reply:
We do `store_deal_state` with a new `left_boundary` for _each_ deal:
1. In the first loop, we store a new `left_boundary` with an updated `log.block_number` if there's one.
2. In the second loop, we're iterating over _each_ deal once again, moving its left boundary to `min(latest_block, change.right_boundary)`

I'm not sure why we do the second for-loop, I haven't changed this code, and why we don't do it immediately, but _each_ deal is updated at the end.

So, I don't understand what you're trying to say. My best guess is you want to do smth like immediately, without waiting for the second loop, right?
```
if change.log != nil:
    store_deal_state(deal_id, change.log.block)
else:
    store_deal_state(deal_id, min(latest_block, change.right_boundary)    
```